marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/matcher-impl/matcher-impl.xqy

```diff
@@ -355,20 +373,84 @@ declare function match-impl:query-map-to-query(
       return (
         if (fn:exists($query-prov)) then
           for $query-hash in document {$query}//schema-element(cts:query) ! xdmp:md5(document{.})
-          where fn:not(map:contains($query-prov,$query-hash)) or ((map:get($query-prov, $query-hash) => map:get("type")) = "reduce")
+          where fn:not(map:contains($query-prov, $query-hash)) or ((map:get($query-prov, $query-hash) => map:get("type")) = "reduce")
           return map:put($query-prov, $query-hash, $sub-query-map)
         else (),
         $query
+      ),
+
+      match-impl:multi-struct-prop-multi-value-queries($query-map, $document, $target-entity)
       )
     (: We want to be certain that didn't lose any queries in the match ruleset, since we don't call the function if no values exist :)
-    where fn:exists($queries) and fn:count($queries) eq fn:count($sub-query-maps)
+    (: FSnow: need to remove the count check since we are returning one query for multiple properties in the structured property case :)
+    where fn:exists($queries) (: and fn:count($queries) eq fn:count($sub-query-maps) :)
     return (
-      map:put($cached-queries, $query-map => map:get("matchRulesetId"), $queries),
+      (: no caching in the structured properties case (yet) :)
+      if (fn:not(map:contains($query-map, "multiStructPropMultiValueMap"))) then
+        map:put($cached-queries, $query-map => map:get("matchRulesetId"), $queries)
+      else
+        ()
+      ,
       $queries
     )
 
 };
 
+declare function match-impl:multi-struct-prop-multi-value-queries($query-map, $document, $target-entity as xs:string?)
+  as cts:query*
+{
+  let $multi-struct-prop-multi-value-map := map:get($query-map, "multiStructPropMultiValueMap")
+  return
+    if (fn:empty($multi-struct-prop-multi-value-map) or fn:empty($target-entity)) then
+      ()
+    else
+      let $is-json := (xdmp:node-kind($document) = "object" or fn:exists($document/(object-node()|array-node())))
+      let $count := fn:count(map:get($query-map, "matchQueries"))
+      let $count := if ($count eq 0) then 1 else $count
+      let $weight := map:get($query-map, "weight") div $count
+
+      let $rev-map := -$multi-struct-prop-multi-value-map
```

Comment:
At first glance, I don't know what the values of the map are, so I don't know what the keys of this map will be. I think it needs a name that defines what it's a map of.